src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -177,155 +204,234 @@ private HashHeightPair TipToChase()
+            if (crossChainTransfers == null)
             {
-                Recipient.FederationWallet wallet = this.federationWalletManager.GetWallet();
+                crossChainTransfers = Get(
+                    this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
+                        this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            }
 
-                using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            var tracker = new StatusChangeTracker();
+            foreach (CrossChainTransfer partialTransfer in crossChainTransfers)
+            {
+                // Verify that the transaction input UTXO's have been reserved by the wallet.
+                if (!SanityCheck(partialTransfer.PartialTransaction, wallet, partialTransfer.Status == CrossChainTransferStatus.FullySigned))
                 {
-                    dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
+                    tracker.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
+                }
+            }
 
-                    CrossChainTransfer[] partialTransfers = this.Get(dbreezeTransaction,
-                        this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
-                            this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            if (tracker.Count == 0)
+                return;
+
+            using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            {
+                dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
 
-                    foreach (CrossChainTransfer partialTransfer in partialTransfers)
+                try
+                {
+                    foreach (KeyValuePair<ICrossChainTransfer, CrossChainTransferStatus?> kv in tracker)
                     {
-                        if (!SanityCheck(partialTransfer.PartialTransaction, wallet))
-                        {
-                            this.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
-                            await this.PutTransferAsync(dbreezeTransaction, partialTransfer);
-                        }
+                        this.PutTransfer(dbreezeTransaction, kv.Key);
                     }
 
                     dbreezeTransaction.Commit();
+
+                    this.UpdateLookups(tracker);
+                }
+                catch (Exception err)
+                {
+                    this.logger.LogError("Error during database update: {0}", err.Message);
+                    this.logger.LogTrace("(-):[SANITY_ERROR]");
+
+                    // Restore expected store state in case the calling code retries / continues using the store.
+                    dbreezeTransaction.Rollback();
+                    throw err;
                 }
```

Comment:
This type of exception is the same in 6 places - apart from the upper case error tag - maybe move into one method ?

Reply:
Added `RollbackAndThrowTransactionError`.

Reply:
Thanks